src/dataflow/src/render/join/mod.rs

```diff
@@ -0,0 +1,275 @@
+                demand.extend(expr.support());
+            }
+        }
+        demand.sort();
+        demand.dedup();
+        // We only want to remove columns that are presented as outputs (i.e. can be found as in
+        // `columns`). Other columns have yet to be introduced, and we shouldn't have any opinion
+        // about them yet.
+        demand.retain(|column| columns.contains_key(column));
+        // Project `before` output columns using current locations of demanded columns.
+        before = before.project(demand.iter().map(|column| columns[column]));
+        // Update `columns` to reflect location of retained columns.
+        columns.clear();
+        for (index, column) in demand.iter().enumerate() {
+            columns.insert(*column, index);
+        }
+
+        // If `mfp` is a permutation of the columns present in `columns`, then we can
+        // apply that permutation to `before` and `columns`, so that `mfp` becomes the
+        // identity operation.
+        if mfp.expressions.is_empty()
+            && mfp.predicates.is_empty()
+            && mfp.projection.len() == columns.len()
+            && mfp.projection.iter().all(|col| columns.contains_key(col))
+        {
+            // The projection we want to apply to `before`  comes to us from `mfp` in the
+            // extended output column reckoning.
+            let projection = mfp
+                .projection
+                .iter()
+                .map(|col| columns[col])
+                .collect::<Vec<_>>();
+            before = before.project(projection);
+            // Update the physical locations of each output column.
+            columns.clear();
+            for (index, column) in mfp.projection.iter().enumerate() {
+                columns.insert(*column, index);
+            }
+        }
+
+        // TODO(mcsherry): perform more optimizations here.
+        before.remove_undemanded();
+
+        // Cons up an instance of the closure with the closed-over state.
+        Self {
+            ready_equivalences,
+            before,
+        }
+    }
+
+    /// True iff the closure neither filters nor transforms recorcds.
+    pub fn is_identity(&self) -> bool {
+        self.ready_equivalences.is_empty() && self.before.is_identity()
+    }
+}
+
+/// Maintained state as we construct join dataflows.
+///
+/// This state primarily tracks the *remaining* work that has not yet been applied to a
+/// stream of partial results.
+///
+/// This state is meant to reconcile the logical operations that remain to apply (e.g.
+/// filtering, expressions, projection) and the physical organization of the current stream
+/// of data, which columns may be partially assembled in non-standard locations and which
+/// may already have been partially subjected to logic we need to apply.
+pub struct JoinBuildState {
+    /// Map from expected locations in extended output column reckoning to physical locations.
+    pub column_map: HashMap<usize, usize>,
```

Comment:
I dont think any of these fields need to be publicly visible (I tried compiling with all 3 `pub`s removed and it worked fine)

Reply:
The different join implementations do need at least `column_map`, for their `permute_map` calls. If they can get in there now it's just because the privacy levels aren't enforced. I'll check it out and see what makes sense.